pkg/vm/vm.go

```diff
@@ -156,22 +177,55 @@ func (v *VM) Reset(t trigger.Type) {
 	v.uncaughtException = nil
 	v.refs = 0
 	v.gasConsumed = 0
-	v.GasLimit = 0
+	v.gasLimit = 0
 	v.SyscallHandler = nil
 	v.LoadToken = nil
 	v.trigger = t
 	v.invTree = nil
 }
 
-// GasConsumed returns the amount of GAS consumed during execution.
+// ExecFeeFactorMultiplier is a multiplier applied to ExecFeeFactor starting
+// from [config.HFFaun] hardfork to provide fractional ExecFeeFactor
+// functionality.
+const ExecFeeFactorMultiplier = 10000
+
+// GasConsumed returns the amount of GAS consumed during execution in Datoshi
+// units rounded from picoGAS to the upper integer.
 func (v *VM) GasConsumed() int64 {
-	return v.gasConsumed
+	return PicoGasToDatoshi(v.gasConsumed)
 }
 
-// AddGas consumes the specified amount of gas. It returns true if gas limit wasn't exceeded.
-func (v *VM) AddGas(gas int64) bool {
+// GasLeft returns the amount of GAS left in Datoshi units rounded from picoGAS
+// to the lower integer (if positive).
+func (v *VM) GasLeft() *big.Int {
+	if v.gasLimit == -1 {
+		return big.NewInt(v.gasLimit)
+	}
+	return big.NewInt((v.gasLimit - v.gasConsumed) / ExecFeeFactorMultiplier)
+}
+
+// PicoGasToDatoshi divides x by ExecFeeFactorMultiplier and rounds the result
+// up to the upper integer in case of non-zero remnant.
+func PicoGasToDatoshi(x int64) int64 {
+	q := x / ExecFeeFactorMultiplier
```

Comment:
`return (x + ExecFeeFactorMultiplier - 1)/ ExecFeeFactorMultiplier`, unnecessary conditions should be avoided.